### ISSUE TITLE
confirm: position_ids's dimension

### ISSUE THREAD

**Issue by jasonzhang761213** [elbruno/ElBruno.QwenTTS#51]
for the onnx module of talker_prefill.onnx
it contains one input named "position_ids"
which dimension was （3，batch ，seqlen)

because I want to try batch inference , but the first "3" make me confused.
and the position_ids fill only 0~seq ,repeat 3times,

**Comment by elbruno** [comment]
Thanks for the question! 

The \position_ids\ tensor uses **M-RoPE (Multimodal Rotary Position Embedding)** with 3 spatial axes, inherited from the upstream Qwen3-TTS architecture. For text-to-speech, all 3 axes hold identical sequential position values—they're not actually treating the sequence as multidimensional, but the shape is preserved to maintain consistency with the multi-modal design.

You can see the implementation in **LanguageModel.cs** (lines 184-187), where the position embeddings are applied. This design choice is documented in our **ARCHITECTURE.md** for reference.

Feel free to reach out if you have more questions about the model structure!

**Comment by elbruno** [comment]
This issue has been resolved in v1.0.1! The M-RoPE position_ids explanation is now part of the release. See: https://github.com/elbruno/ElBruno.QwenTTS/releases/tag/v1.0.1

**Comment by jasonzhang761213** [comment]
I believed this tensor useful, but just dont know how to make batch more than 1.

for example, if batch =1,seq =5
the position_ids should be
[
    [[1, 2, 3, 4, 5]],
    [[1, 2, 3, 4, 5]],
    [[1, 2, 3, 4, 5]]
]

but how about batch=2 ，seq =5 ， will the tensor turn to be？
[
    [[1, 2, 3, 4, 5],  [1, 2, 3, 4, 5]],
    [[1, 2, 3, 4, 5],  [1, 2, 3, 4, 5]],
    [[1, 2, 3, 4, 5],  [1, 2, 3, 4, 5]],
]

**Comment by elbruno** [comment]
Hi @jasonzhang761213! 👋

Great question about batch inference with position_ids! Yes, your understanding is **100% correct**. Let me confirm the behavior:

## M-RoPE Position IDs for Batch Inference

For **batch=1, seq=5**, the shape is (3, 1, 5):
```
[
    [[0, 1, 2, 3, 4]],
    [[0, 1, 2, 3, 4]],
    [[0, 1, 2, 3, 4]]
]
```
_(Note: positions start at 0, not 1)_

For **batch=2, seq=5**, the shape becomes (3, 2, 5):
```
[
    [[0, 1, 2, 3, 4],  [0, 1, 2, 3, 4]],
    [[0, 1, 2, 3, 4],  [0, 1, 2, 3, 4]],
    [[0, 1, 2, 3, 4],  [0, 1, 2, 3, 4]]
]
```

## Code Example for Batch=2

Here's how you'd construct it in C#:

```csharp
int batchSize = 2;
int seqLen = 5;
var positionIds = new long[3, batchSize, seqLen];

for (int ax = 0; ax < 3; ax++)
    for (int b = 0; b < batchSize; b++)
        for (int i = 0; i < seqLen; i++)
            positionIds[ax, b, i] = i;
```

## How Broadcasting Works

The **3 spatial axes** come from M-RoPE (Multimodal Rotary Position Embedding). Each axis gets identical position values because:
1. The ONNX model's RoPE implementation expects shape (3, batch, seq)
2. Each batch item gets its **own independent position sequence** (both start at 0)
3. All 3 M-RoPE axes use the **same position values** — they're not treating the sequence as multidimensional, it's just the architecture design

During decode steps, each batch item's position increments independently based on its own sequence length. You can see this in [LanguageModel.cs lines 382-384](https://github.com/elbruno/ElBruno.QwenTTS/blob/main/src/ElBruno.QwenTTS.Core/Models/LanguageModel.cs#L382-L384) where decode positions are computed.

## Key Insight for Batch Implementation

When implementing batch inference, make sure to also update:
- **attention_mask**: shape (batch, seq) — all 1s for valid tokens
- **past_keys/past_values**: shape (layers, batch, kv_heads, past_seq, head_dim)
- **inputs_embeds**: shape (batch, seq, hidden_size)

The current implementation hardcodes batch=1, but the pattern you've identified is exactly right for extending it to batch>1! 🚀

Hope this helps with your batch inference implementation! Let me know if you have any other questions.

**Comment by elbruno** [comment]
All questions about position_ids and batch inference have been answered with code examples and implementation guidance. Feel free to reopen if you have additional questions!
